alibi/explainers/shap_wrappers.py

```diff
@@ -0,0 +1,1431 @@
+            model outputs.
+        expected_value:
+            A list containing the expected value of the prediction for each class. Its length should be equal to that of
+            `shap_output`.
+
+        Returns
+        -------
+            An explanation object containing the shap values and prediction in the `data` field, along with a `meta`
+            field containing additional data. See usage examples in the method overview for details.
+        """
+        y = kwargs.get('y')
+        if not y:
+            y = np.array([])
+        cat_vars_start_idx = kwargs.get('cat_vars_start_idx', ())  # type: Tuple[int]
+        cat_vars_enc_dim = kwargs.get('cat_vars_enc_dim', ())  # type: Tuple[int]
+        summarise_result = kwargs.get('summarise_result', False)  # type: bool
+
+        # check if interactions were computed
+        if len(shap_output[0].shape) == 3:
+            shap_interaction_values = shap_output
+            # shap values are the sum over all shap interaction values for each instance
+            shap_values = [interactions.sum(axis=2) for interactions in shap_output]
+        else:
+            shap_interaction_values = [np.array([])]
+            shap_values = shap_output
+
+        self._check_result_summarisation(summarise_result, cat_vars_start_idx, cat_vars_enc_dim)
+        if self.summarise_result:
+            summarised_shap = []
+            for shap_array in shap_values:
+                summarised_shap.append(sum_categories(shap_array, cat_vars_start_idx, cat_vars_enc_dim))
+            shap_values = summarised_shap
+            if shap_interaction_values[0].size != 0:
+                summarised_shap_interactions = []
+                for shap_array in shap_interaction_values:
+                    summarised_shap_interactions.append(
+                        sum_categories(shap_array, cat_vars_start_idx, cat_vars_enc_dim)
+                    )
+                shap_interaction_values = summarised_shap_interactions
+
+        raw_predictions, loss = [], []
+        # raw output of a regression or classification task. Will not work for pyspark.
+        if self.model_output != 'log_loss':
+            raw_predictions = self._explainer.model.predict(X, tree_limit=self.tree_limit)
+        # loss function values explained
+        if y is not None or self.model_output == 'log_loss':
+            loss = self._explainer.model.predict(X, y, tree_limit=self.tree_limit)
+        # predicted class
+        if self.task != 'regression' and raw_predictions:
+            argmax_pred = np.argmax(np.atleast_2d(raw_predictions), axis=1)
+        else:
+            argmax_pred = []
+        importances = rank_by_importance(shap_values, feature_names=self.feature_names)
+
+        if self._explainer.model.model_type == 'catboost':
+            if not (isinstance(X, np.ndarray) or isinstance(X, pd.DataFrame)):
```

Comment:
Would `if isinstance(X, catboost.Pool)` not work here? Assuming `X` really is of type `catboost.Pool`, the type information should be available in the user's Python session.

Reply:
I was mainly thinking of someone who uses alibi and doesn't want to install `catboost`?

Reply:
This shouldn't matter, I'm not proposing to take on a dependency, but if the `if` statement happens to be true, surely `catboost.Pool` exists in the Python session already?